### PR TITLE
Track id label fixes

### LINKED ISSUE #1134: Remote Track ID and Label are blank for answers using Unified Plan, and sometimes after renegotiation
I have a fix ready and am going to open a pull request associated with this issue. Would greatly appreciate feedback!

### Your environment.
  - Version: 364d0de
  - Browser: Firefox and Chrome
  - Other information: please see below

### What did you do?

This happens in two cases:

1. I initiated a peer connection from server to Chrome or Firefox. Both of them default to unified plan and no longer return the track ID and label in the `a=ssrc:<ssrc> msid:<track_id> <track_label>` format, but in `a=msid:<track_id> <track_label>` format.

Should be fixed by: b078d57, already in #1135.

2.
    - I created a server with a `sendrecv` video transceiver
    - The server created an offer with a`sendrecv` media section.
    - The browser answered with a `recvonly` media section.
    - The browser called `addTrack()` and requested negotiation from the server. 
    - The server created an offer 
    - The browser answered with a `sendrecv` media section and track ID / label
    - The `OnTrack` handler was called with the right `ssrc`, but blank `id` and `label` fields.

Should be fixed by 15beeb1, already in #1135.

### What did you expect?

I expected `track.ID()`  and `track.Label()` to be set, but they were blank: `""` in both cases.

### What happened?

There are two reasons why this happens:

#### 1. Unified Plan uses a new attribute key: `a=ssrc` 

From [Unified Plan Transition Guide](https://webrtc.org/getting-started/unified-plan-transition-guide):

> People who do detailed SDP parsing and care about msid attributes will have to check that their parsing code picks up the new format (a=msid)

The current code tries to extract it from:

```
a=ssrc:<ssrc> msid:<msid> <label>
```

However, both Firefox and Chrome now send it as:

```
a=msid:<msid> <label>
a=ssrc:<ssrc> cname:<cname>
```

#### 2. Track.ID and Track.Label are not sent in the first SDP, but only during the renegotiation in the second answer

The pion/webrtc server generates an offer:

> m=video 9 UDP/TLS/RTP/SAVPF 96 98 102
> c=IN IP4 0.0.0.0
> a=setup:actpass
> **a=mid:0**
> a=ice-ufrag:CdEJRWUFkiQvhkHz
> a=ice-pwd:UulhpVNabGWvAHoAFhCbQNQYLjqmBOFR
> a=rtcp-mux
> a=rtcp-rsize
> a=rtpmap:96 VP8/90000
> a=rtpmap:98 VP9/90000
> a=rtpmap:102 H264/90000
> a=fmtp:102 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42001f
> a=ssrc:2596996162 cname:TjHVWxkKPtxTAeFD
> a=ssrc:2596996162 msid:TjHVWxkKPtxTAeFD NjPSPGQVICwMHQxo
> a=ssrc:2596996162 mslabel:TjHVWxkKPtxTAeFD
> a=ssrc:2596996162 label:NjPSPGQVICwMHQxo
> a=msid:TjHVWxkKPtxTAeFD NjPSPGQVICwMHQxo
> **a=sendrecv**
> ... candidates and other streams

The first answer from the browser (`addTrack` hasn't been called yet):

> m=video 56812 UDP/TLS/RTP/SAVPF 96 98
> **a=recvonly**
> a=end-of-candidates
> a=fmtp:96 max-fs=12288;max-fr=60
> a=fmtp:98 max-fs=12288;max-fr=60
> a=ice-pwd:a1dbf2ff4fb7ad70afd549ecfc743786
> a=ice-ufrag:91ad396c
> **a=mid:0**
> a=rtcp-mux
> a=rtpmap:96 VP8/90000
> a=rtpmap:98 VP9/90000
> a=setup:active
> **a=ssrc:4199521927 cname:{f97c2d01-98b1-d848-9a6e-5786ee794fff}**
> m=audio 56812 UDP/TLS/RTP/SAVPF 111 0 8 9

At this point, a new `*webrtc.Track` is created in `PeerConnection.startReceiver()`, with `id=""` and `label=""`. The `startReceiver` blocks in `determinePayloadType()` and no packets have been sent because the `addTrack` hasn't been called in the browser yet.

Next, a track is added in the browser, and the browser requests negotiation from the server. Below is a second offer from the pion/webrtc server:

> **a=mid:0**
> a=rtcp-mux
> a=rtcp-rsize
> a=rtpmap:96 VP8/90000
> a=rtpmap:98 VP9/90000
> a=rtpmap:102 H264/90000
> a=fmtp:102 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42001f
> a=ssrc:2596996162 cname:TjHVWxkKPtxTAeFD
> a=ssrc:2596996162 msid:TjHVWxkKPtxTAeFD NjPSPGQVICwMHQxo
> a=ssrc:2596996162 mslabel:TjHVWxkKPtxTAeFD
> a=ssrc:2596996162 label:NjPSPGQVICwMHQxo
> a=msid:TjHVWxkKPtxTAeFD NjPSPGQVICwMHQxo
> **a=sendrecv**
> ... candidates

And the browser's final answer is:

> m=video 56812 UDP/TLS/RTP/SAVPF 96 98
> **a=sendrecv**
> a=end-of-candidates
> a=fmtp:96 max-fs=12288;max-fr=60
> a=fmtp:98 max-fs=12288;max-fr=60
> a=ice-pwd:a1dbf2ff4fb7ad70afd549ecfc743786
> a=ice-ufrag:91ad396c
> **a=mid:0**
> **a=msid:{dac0424f-0b68-bd40-9e47-04780b0c4f70} {9d8ee2b5-345e-a041-92f3-8851865df6b3}**
> a=rtcp-mux
> a=rtpmap:96 VP8/90000
> a=rtpmap:98 VP9/90000
> a=setup:active
> **a=ssrc:4199521927 cname:{f97c2d01-98b1-d848-9a6e-5786ee794fff}**

After successful renegotiation, the `OnTrack` handler is called with `track.ID() == ""` and `track.Label() == ""`, because those were the values during the first negotiation.

### PR DESCRIPTION
**Pull request by jeremija**
#### Description
See #1134

#### Reference issue
Fixes #1134

### CODE REVIEW

File: peerconnection.go [hunk 1]
```diff
@@ -784,6 +784,13 @@ func (pc *PeerConnection) startReceiver(incoming trackDetails, receiver *RTPRece
 		return
 	}
 
+	// set track id and label early so they can be set as new track information
+	// is received from the SDP.
+	receiver.Track().mu.Lock()
+	receiver.Track().id = incoming.id
+	receiver.Track().label = incoming.label
+	receiver.Track().mu.Unlock()
```
Comment (at-wat) [inline peerconnection.go:792]: I think it is not guaranteed to be executed before the expected timing.
```diff
$ git diff -w
diff --git a/peerconnection.go b/peerconnection.go
index 2c3c69c..d1c6b4c 100644
--- a/peerconnection.go
+++ b/peerconnection.go
@@ -791,6 +791,7 @@ func (pc *PeerConnection) startReceiver(incoming trackDetails, receiver *RTPRece
        receiver.Track().label = incoming.label
        receiver.Track().mu.Unlock()
 
+       go func() {
                if err = receiver.Track().determinePayloadType(); err != nil {
                        pc.log.Warnf("Could not determine PayloadType for SSRC %d", receiver.Track().SSRC())
                        return
@@ -820,6 +821,7 @@ func (pc *PeerConnection) startReceiver(incoming trackDetails, receiver *RTPRece
                } else {
                        pc.log.Warnf("OnTrack unset, unable to handle incoming media streams")
                }
+       }()
 }
 
 // startRTPReceivers opens knows inbound SRTP streams from the RemoteDescription
@@ -858,7 +860,7 @@ func (pc *PeerConnection) startRTPReceivers(incomingTracks map[uint32]trackDetai
 
                        delete(incomingTracks, ssrc)
                        localTransceivers = append(localTransceivers[:i], localTransceivers[i+1:]...)
-                       go pc.startReceiver(incoming, t.Receiver())
+                       pc.startReceiver(incoming, t.Receiver())
                        break
                }
        }
@@ -872,7 +874,7 @@ func (pc *PeerConnection) startRTPReceivers(incomingTracks map[uint32]trackDetai
                                pc.log.Warnf("Could not add transceiver for remote SSRC %d: %s", ssrc, err)
                                continue
                        }
-                       go pc.startReceiver(incoming, t.Receiver())
+                       pc.startReceiver(incoming, t.Receiver())
                }
        }
 }
@@ -960,7 +962,7 @@ func (pc *PeerConnection) drainSRTP() {
                                        pc.log.Warnf("Could not add transceiver for remote SSRC %d: %s", ssrc, err)
                                        return false
                                }
-                               go pc.startReceiver(incoming, t.Receiver())
+                               pc.startReceiver(incoming, t.Receiver())
                                return true
                        }
                }
```
would be safe.

File: sdp.go [hunk 2]
```diff
@@ -55,6 +57,16 @@ func trackDetailsFromSDP(log logging.LeveledLogger, s *sdp.SessionDescription) m
...
+			// TODO make "msid" a constant
+			if attr.Key == "msid" {
+				split := strings.Split(attr.Value, " ")
+				if len(split) == 2 {
+					trackLabel = split[0]
+					trackID = split[1]
+				}
+			}
+
 			if attr.Key == sdp.AttrKeySSRC {
```
Comment (at-wat) [inline sdp.go:70]:
```go
switch attr.Key {
```
would be better.

File: sdp.go [hunk 3]
```diff
@@ -55,6 +57,16 @@ func trackDetailsFromSDP(log logging.LeveledLogger, s *sdp.SessionDescription) m
 				}
 			}
 
+			// Handle a=msid for Unified plan
+			// TODO make "msid" a constant
```
Comment (at-wat) [inline sdp.go:61]: https://github.com/pion/sdp/blob/99a4a121c465b7ab633505dcea9a96e2ccca9066/jsep.go#L11-L22

Comment (jeremija) [reply sdp.go:61]: Thanks! I created a pull request in `pion/sdp`: https://github.com/pion/sdp/pull/47

File: sdp.go (hunk 2 above)
Comment (jeremija) [reply sdp.go:70]: Thanks, I can modify it to use `switch()`. I'm just not sure if there are any guarantees that `a=msid` section will __always__ come before `a=ssrc`. If there are no guarantees, then we might have to keep a reference to last `trackDetails`, and update it when the `a=msid` section for the same track comes after `a=ssrc`.

**Comment by jeremija** [comment]
Thanks for your comments, @at-wat! 

I believe I've addressed all of your change requests.

Copying my comment from one of the code comment threads I'm about to mark as resolved:

> I'm just not sure if there are any guarantees that a=msid section will always come before a=ssrc. If there are no guarantees, then we might have to keep a reference to last trackDetails, and update it when the a=msid section for the same track comes after a=ssrc.

File: sdp.go [hunk 4]
```diff
@@ -53,9 +56,17 @@ func trackDetailsFromSDP(log logging.LeveledLogger, s *sdp.SessionDescription) m
 						delete(incomingTracks, uint32(rtxRepairFlow)) // Remove if rtx was added as track before
 					}
 				}
-			}
 
-			if attr.Key == sdp.AttrKeySSRC {
+			// Handle a=msid for Unified plan
+			// TODO make "msid" a constant
```
Comment (at-wat) [inline sdp.go:61]: this TODO comment is no longer needed

**Comment by Sean-Der** [comment]
Sorry for the slow review everyone :( honestly I am not sure either! I will dig now, but IMO it adds tests (and doesn't break any existing ones) so it is a merge from me!

**Comment by Sean-Der** [comment]
IMO we really need to make the Pion media API match the browser. A lot of this confusion comes from the fact that we can't run our WASM tests against the media API.

A lot of bike shedding happened around DataChannels before the WASM stuff, and it all went away.

**Comment by jeremija** [comment]
@at-wat & @Sean-Der  thanks so much for your help! I'm going to merge this and then I can push a fix if we realize that `a=msid` sometimes comes after `a=ssrc`.

**Pull request merged**